File: backend/src/main/java/com/built4u/pos/docsettings/DocSettings.java
package com.built4u.pos.docsettings;

import com.built4u.pos.common.tenant.YesNoConverter;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

/**
 * Per-site document/branding config ({@code pos_doc_settings}), keyed by site_id.
 * Consumed by the report-PDF letterhead and the sale-receipt PDF. No row for a
 * site ⇒ the service falls through to hard-coded defaults.
 */
@Entity
@Table(name = "pos_doc_settings")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class DocSettings {

    @Id
    @Column(name = "site_id", nullable = false)
    private Long siteId;

    @Column(name = "business_name", length = 150)
    private String businessName;

    @Column(name = "address_line", length = 255)
    private String addressLine;

    @Column(name = "contact_line", length = 150)
    private String contactLine;

    @Column(length = 40)
    private String tin;

    @Column(name = "footer_note", length = 255)
    private String footerNote;

    @Column(name = "accent_color", nullable = false, length = 7)
    @Builder.Default
    private String accentColor = "#1D4ED8";

    @Column(name = "receipt_title", nullable = false, length = 60)
    @Builder.Default
    private String receiptTitle = "SALES RECEIPT";

    @Column(name = "receipt_footer", nullable = false, length = 255)
    @Builder.Default
    private String receiptFooter = "Thank you!";

    // ── Logo (shared by report PDFs and receipts) ────────────────────────────
    @Lob
    @Column(name = "logo_image")
    private byte[] logoImage;

    @Column(name = "logo_mime", length = 40)
    private String logoMime;

    @Column(name = "logo_position", nullable = false, length = 10)
    @Builder.Default
    private String logoPosition = "LEFT";

    // ── Report-PDF page setup ────────────────────────────────────────────────
    @Column(name = "show_logo_pdf", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean showLogoPdf = true;

    @Column(name = "paper_size", nullable = false, length = 10)
    @Builder.Default
    private String paperSize = "A4";

    @Column(name = "orientation", nullable = false, length = 10)
    @Builder.Default
    private String orientation = "LANDSCAPE";

    @Column(name = "margin_preset", nullable = false, length = 10)
    @Builder.Default
    private String marginPreset = "NORMAL";

    @Column(name = "font_scale", nullable = false, length = 10)
    @Builder.Default
    private String fontScale = "NORMAL";

    @Column(name = "zebra_striping", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean zebraStriping = true;

    @Column(name = "show_page_numbers", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean showPageNumbers = true;

    @Column(name = "show_timestamp", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean showTimestamp = true;

    @Column(name = "show_printed_by", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean showPrintedBy = true;

    // ── Receipt customization ────────────────────────────────────────────────
    @Column(name = "show_logo_receipt", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean showLogoReceipt = false;

    @Column(name = "receipt_header_note", length = 255)
    private String receiptHeaderNote;

    @Column(name = "receipt_show_cashier", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean receiptShowCashier = true;

    @Column(name = "receipt_show_customer", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean receiptShowCustomer = true;

    @Column(name = "receipt_show_voucher", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean receiptShowVoucher = true;

    /** Physical receipt layout: THERMAL_80MM (default) or BOND_LETTER (US Letter). */
    @Column(name = "receipt_format", nullable = false, length = 20)
    @Builder.Default
    private String receiptFormat = "THERMAL_80MM";

    // ── Network receipt printer + cash drawer (per site) ─────────────────────
    /** LAN printer host/IP; blank ⇒ network printing unavailable for this site. */
    @Column(name = "receipt_printer_host", length = 100)
    private String receiptPrinterHost;

    @Column(name = "receipt_printer_port", nullable = false)
    @Builder.Default
    private Integer receiptPrinterPort = 9100;

    @Column(name = "receipt_printer_enabled", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean receiptPrinterEnabled = false;

    /** Kick the cash drawer when a sale receipt prints. */
    @Column(name = "open_drawer_on_sale", nullable = false, length = 1)
    @Convert(converter = YesNoConverter.class)
    @Builder.Default
    private Boolean openDrawerOnSale = false;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "updated_by", length = 50)
    private String updatedBy;
}

File: frontend/src/lib/docsettings.ts
import api from '@/lib/api'
import type { AxiosError } from 'axios'
import * as agent from '@/lib/printagent'

export const docErr = (e: unknown, f: string) =>
  (e as AxiosError<{ message?: string }>).response?.data?.message ?? f

export interface DocSettings {
  siteId: number
  usingDefault: boolean
  businessName: string | null
  addressLine: string | null
  contactLine: string | null
  tin: string | null
  footerNote: string | null
  accentColor: string
  receiptTitle: string
  receiptFooter: string
  // Logo + report-PDF page setup
  hasLogo: boolean
  logoMime: string | null
  logoPosition: string
  showLogoPdf: boolean
  paperSize: string
  orientation: string
  marginPreset: string
  fontScale: string
  zebraStriping: boolean
  showPageNumbers: boolean
  showTimestamp: boolean
  showPrintedBy: boolean
  // Receipt customization
  showLogoReceipt: boolean
  receiptHeaderNote: string | null
  receiptShowCashier: boolean
  receiptShowCustomer: boolean
  receiptShowVoucher: boolean
  receiptFormat: string
  // Network printer + drawer
  receiptPrinterHost: string | null
  receiptPrinterPort: number
  receiptPrinterEnabled: boolean
  openDrawerOnSale: boolean
  updatedBy: string | null
  updatedAt: string | null
}
export interface SaveDocSettings {
  businessName?: string
  addressLine?: string
  contactLine?: string
  tin?: string
  footerNote?: string
  accentColor?: string
  receiptTitle?: string
  receiptFooter?: string
  logoPosition?: string
  showLogoPdf?: boolean
  paperSize?: string
  orientation?: string
  marginPreset?: string
  fontScale?: string
  zebraStriping?: boolean
  showPageNumbers?: boolean
  showTimestamp?: boolean
  showPrintedBy?: boolean
  showLogoReceipt?: boolean
  receiptHeaderNote?: string
  receiptShowCashier?: boolean
  receiptShowCustomer?: boolean
  receiptShowVoucher?: boolean
  receiptFormat?: string
  // Network printer + drawer
  receiptPrinterHost?: string
  receiptPrinterPort?: number
  receiptPrinterEnabled?: boolean
  openDrawerOnSale?: boolean
}

export async function getDocSettings(): Promise<DocSettings> {
  const { data } = await api.get<DocSettings>('/admin/doc-settings')
  return data
}
/** Business identity + logo placement (module DOC_SETTINGS). */
export async function saveDocIdentity(body: SaveDocSettings): Promise<DocSettings> {
  const { data } = await api.put<DocSettings>('/admin/doc-settings/identity', body)
  return data
}
/** Report-PDF layout (module PDF_CONFIG). */
export async function saveDocPdf(body: SaveDocSettings): Promise<DocSettings> {
  const { data } = await api.put<DocSettings>('/admin/doc-settings/pdf', body)
  return data
}
/** Sale-receipt customization (module RECEIPT_CONFIG). */
export async function saveDocReceipt(body: SaveDocSettings): Promise<DocSettings> {
  const { data } = await api.put<DocSettings>('/admin/doc-settings/receipt', body)
  return data
}

/** Upload a logo (PNG/JPEG, ≤512 KB) used on report PDFs and receipts. */
export async function uploadDocLogo(file: File): Promise<DocSettings> {
  const form = new FormData()
  form.append('file', file)
  const { data } = await api.post<DocSettings>('/admin/doc-settings/logo', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}
export async function deleteDocLogo(): Promise<DocSettings> {
  const { data } = await api.delete<DocSettings>('/admin/doc-settings/logo')
  return data
}
/** Cache-busting URL for the current logo (auth cookie/header applied by <img> via api base is N/A — use fetch). */
export async function fetchDocLogoObjectUrl(): Promise<string | null> {
  try {
    const res = await api.get('/admin/doc-settings/logo', { responseType: 'blob' })
    return URL.createObjectURL(res.data as Blob)
  } catch {
    return null
  }
}

// Printer actions go through the print-agent helper: it relays via a local agent
// when one is running (needed when the backend is hosted off the shop's LAN) and
// otherwise prints straight from the server, exactly as before.

/** Print a sale receipt to the site's network thermal printer (opens the drawer if configured). */
export async function printSaleReceipt(salesNumber: string): Promise<void> {
  await agent.printReceipt(salesNumber)
}
/** Print a test slip + kick the drawer to confirm the printer is reachable. */
export async function testPrinter(): Promise<void> {
  await agent.printTest()
}
/** Open the cash drawer without printing. */
export async function openDrawer(): Promise<void> {
  await agent.openDrawer()
}

/** Fetch a sale's receipt PDF (with auth) and open it in a new tab. */
export async function openReceipt(salesNumber: string): Promise<void> {
  const res = await api.get(`/sales/${salesNumber}/receipt`, { responseType: 'blob' })
  const url = URL.createObjectURL(new Blob([res.data], { type: 'application/pdf' }))
  window.open(url, '_blank')
  setTimeout(() => URL.revokeObjectURL(url), 60_000)
}
